### DISCUSSION TITLE
RouterOS Serializer

### DISCUSSION THREAD

**Discussion by aascedu** [amyasnikov/validity#119] [Q&A]
Hey guys, I was trying to add a Mikrotik Serializer using RouterOS as extraction method and I get this in my Netbox logs :

`10.200.84.112 - - [25/Sep/2024:11:58:25 +0000] "POST /plugins/validity/serializers/2/edit/ HTTP/1.1" 302 0 "http://netbox.ipline.bur:8000/plugins/validity/serializers/2/edit/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:130.0) Gecko/20100101 Firefox/130.0"
Serializer Mikrotik Serializer has no template and no Data File defined`

When I try to see the Serialized State of a device : 

![image](https://github.com/user-attachments/assets/d3b73ed2-74ef-43a0-869a-65c23c69e2cc)

My device is from a Data Source, that works for Cisco Device, with my own TTP Template.
Here is more info on the data source : 
![image](https://github.com/user-attachments/assets/366bc45f-77db-40f9-ba99-bcb7ce25e571)

These are the fields I get when trying to add a RouterOS Serializer for Mikrotik equipment :
![image](https://github.com/user-attachments/assets/d2cea905-28f6-46a5-86a0-73a9f7546154)

Since there is not many errors popping off and not many debug info, I am coming here for some help, also I don't know if my problems come from the compatibility issues or if it is me who is not understanding the process and way things work ! 
**_ALSO_** I tested with a TTP Serializer on the same file and it does work well...
And obviously my config is in the correct format because it is exported from the router itself.

Thank you so much in advance if you have some tips to give me !

Arthur.

**Comment by amyasnikov** [comment]
Hey @aascedu, looks like your serializer (Mikrotik Serializer) is not bound to the device (mikrotik-lbr009). You can bind it at individual device level, as well as at Device Type or Manufacturer levels

**Comment by aascedu** [comment]
Hey @amyasnikov 
I did bind it, I tried the 3 different ways to bind it,but nothing.
As I said, I know everything else is setup well because I tried with a custom made TTP Serializer, and I got a Serialized State.
Any clues ?

Thank you ;)

**Comment by aascedu** [comment]
Hello Anton,
I made it work,

In my Mikrotik config, I suspected that some \ char made the Python parser unhappy, so I removed them.
And it was that. Do you know why ? Do you know if it is possible to make it work with backslashes ???

Anyways, thank you for your help. I will probably be asking a lot of questions for the next 3-5 months since I am working with Validity everyday ! ❤️

**Comment by amyasnikov** [comment]
@aascedu do you have an example of the config with these backslashes? Few lines will be enough

**Reply by aascedu** [reply]
@amyasnikov Yes absolutly here are the two elements I changed : 
`set note="*********************************\
    \nADMINISTRATION\
    \nFor further information\
    \nMail : support.entreprises@xxxx.xxx\
    \nPhone : +33333333333\
    \n*********************************" show-at-login=no`
which I changed to `set note="*********************************" show-at-login=no`
and
`add as=xxxxx connect=yes disabled=no hold-time=20s keepalive-time=5s listen=yes local.address=x.x.x.x .role=ebgp multihop=yes name=NAME output.filter-chain=LAN .keep-sent-attributes=yes \
    .network=bgp-network remote.address=x.x.x.x .as=65535 router-id=x.x.x.x`
which I changed to : 
`add as=xxxxx connect=yes disabled=no hold-time=20s keepalive-time=5s listen=yes local.address=x.x.x.x .role=ebgp multihop=yes name=NAME output.filter-chain=LAN .keep-sent-attributes=yes     .network=bgp-network remote.address=x.x.x.x .as=65535 router-id=x.x.x.x`

Would that be why it didn't work ?

**Reply by amyasnikov** [reply]
Probably yes, I have to check it first. As I see there are 2 different cases: backslash inside quotes and and outside of them. Perhaps only one of the cases is causing this problem.

**Comment by aascedu** [comment]
I can check both cases and come back to you!

**Reply by aascedu** [reply]
I am a bit late but here is a part of the config I found to be annoying RouterOS parser :
```
set note="*********************************\
    \nIPLINE ADMINISTRATION\
    \nFor further information\
    \nMail : support.entreprises@dstny.fr\
    \nPhone : +334 27 70 90 00\
    \n*********************************" show-at-login=no
```

will cause that error :
![image](https://github.com/user-attachments/assets/24556727-eb0c-4d7f-ac29-455b2c9e385d)

but when changed to this (without \ or \n) : 
```
set note="*********************************" show-at-login=no
```

![image](https://github.com/user-attachments/assets/2f5cc682-4b2e-44a2-afba-43849843a952)

**Reply by amyasnikov** [reply]
Do you see exactly this one (attention to special characters) when browsing the file in the data source?
```
set note="*********************************\
    \nIPLINE ADMINISTRATION\
    \nFor further information\
    \nMail : support.entreprises@dstny.fr\
    \nPhone : +334 27 70 90 00\
    \n*********************************" show-at-login=no
```
I'm asking because I've tried it and the parser works well

**Reply by aascedu** [reply]
Hey @amyasnikov, to address this, I think what the parser doesn't like is whitespaces, for exemple in my config : 
```
add chain=LAN disabled=no rule=\
    "if (dst in bgp-network) {set bgp-path-prepend 1; accept}"
```
 doesn't work because in the config file, there is a `\` and MOST IMPORTANTLY, a new line  just after `rule=` so it thinks there is multiple whitespaces after... and probably doesn't accept it !
Do you think it would be fixable ?

**Reply by amyasnikov** [reply]
Unfortunately, just copy-pasting from here doesn't help to reproduce the issue. The parser still works fine with the line you provided.

You can do the following to make it reproducible:
1. Create public repository (e.g. on github)
2. Place your full (or at least those parts which in your opinion cause errors) config file there
3. Use this repository as a data source for device configurations in netbox.
4.  Make sure you still have the error with config parsing usgin this repository.
5.  Share the repository with me, so I will be able to perform steps 3 and 4 on my own.

**Comment by aascedu** [comment]
Il check again

**Comment by aascedu** [comment]
I will tag you whenever I have done that! Thank you !

**Reply by amyasnikov** [reply]
@aascedu please create a separate **issue** in case of successful reproduction of the error.

I'm closing this discussion because it
1) has the answer
2) has gone far away from the initial question
